File: web/src/lib/pprof.ts
import { api, getAuthHeader } from "./api";

export type PprofConfig = {
  enabled: boolean;
  authRequired?: boolean;
  continuousEnabled?: boolean;
  continuousIntervalSecs?: number;
  continuousCpuSliceSecs?: number;
  profiles?: string[];
  maxCpuSeconds?: number;
};

export type PprofRuntimeStats = {
  fetchedAt: string;
  goroutines: number;
  memory: {
    alloc: number;
    totalAlloc: number;
    sys: number;
    heapAlloc: number;
    heapInuse: number;
    heapObjects: number;
    numGc: number;
  };
};

export type PprofProfileEntry = {
  name: string;
  flat: number;
  flatPercent: number;
  cum: number;
  cumPercent: number;
};

export type PprofProfileSummary = {
  fetchedAt: string;
  profileType: string;
  totalSamples?: number;
  durationSecs?: number;
  entries: PprofProfileEntry[];
};

export type PprofContinuousSnapshot = {
  fetchedAt: string;
  intervalSecs: number;
  cpuSliceSecs: number;
  profiles: Record<string, PprofProfileSummary>;
  runtimeHistory: PprofRuntimeStats[];
};

export function fetchPprofConfig() {
  return api<PprofConfig>("/api/v1/pprof/config");
}

export function fetchPprofRuntime() {
  return api<PprofRuntimeStats>("/api/v1/pprof/runtime");
}

export function fetchPprofContinuous() {
  return api<PprofContinuousSnapshot>("/api/v1/pprof/continuous");
}

export function fetchPprofProfileSummary(profile: string, seconds?: number, continuous = false) {
  const params = new URLSearchParams();
  if (profile === "cpu" && seconds) {
    params.set("seconds", String(seconds));
  }
  if (continuous) {
    params.set("source", "continuous");
  }
  const query = params.toString();
  return api<PprofProfileSummary>(`/api/v1/pprof/profile/${profile}${query ? `?${query}` : ""}`);
}

export function pprofProfileDownloadUrl(profile: string, seconds?: number) {
  const params = new URLSearchParams();
  if (profile === "cpu" && seconds) {
    params.set("seconds", String(seconds));
  }
  const query = params.toString();
  return `/api/v1/pprof/profile/${profile}/download${query ? `?${query}` : ""}`;
}

export function pprofDebugIndexUrl() {
  return "/api/v1/pprof/config";
}

export function formatPprofBytes(value: number) {
  if (value < 1024) return `${value} B`;
  if (value < 1024 * 1024) return `${(value / 1024).toFixed(1)} KB`;
  if (value < 1024 * 1024 * 1024) return `${(value / (1024 * 1024)).toFixed(1)} MB`;
  return `${(value / (1024 * 1024 * 1024)).toFixed(2)} GB`;
}

export async function downloadPprofProfile(profile: string, seconds?: number) {
  const url = pprofProfileDownloadUrl(profile, seconds);
  const headers = new Headers();
  const auth = getAuthHeader();
  if (auth) {
    headers.set("Authorization", auth);
  }
  const response = await fetch(url, { headers, credentials: "include" });
  if (!response.ok) {
    const body = await response.json().catch(() => ({}));
    throw new Error(body.error ?? `Download failed (${response.status})`);
  }
  const blob = await response.blob();
  const filename =
    response.headers.get("Content-Disposition")?.match(/filename="?([^";]+)"?/)?.[1] ??
    (profile === "cpu" ? "cpu.pprof" : `${profile}.pb.gz`);
  const objectUrl = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = objectUrl;
  anchor.download = filename;
  anchor.click();
  URL.revokeObjectURL(objectUrl);
}

File: internal/domain/pprof.go
package domain

import "time"

type PprofConfig struct {
	Profiles           []string `json:"profiles"`
	MaxCPUSecs         int      `json:"maxCpuSeconds"`
	ContinuousInterval int      `json:"continuousIntervalSecs"`
	ContinuousCPUSlice int      `json:"continuousCpuSliceSecs"`
	Enabled            bool     `json:"enabled"`
	AuthRequired       bool     `json:"authRequired"`
	ContinuousEnabled  bool     `json:"continuousEnabled"`
}

type PprofContinuousSnapshot struct {
	FetchedAt      time.Time                      `json:"fetchedAt"`
	Profiles       map[string]PprofProfileSummary `json:"profiles"`
	RuntimeHistory []PprofRuntimeStats            `json:"runtimeHistory"`
	IntervalSecs   int                            `json:"intervalSecs"`
	CPUSliceSecs   int                            `json:"cpuSliceSecs"`
}

type PprofRuntimeStats struct {
	FetchedAt  time.Time        `json:"fetchedAt"`
	Goroutines int              `json:"goroutines"`
	Memory     PprofMemoryStats `json:"memory"`
}

type PprofMemoryStats struct {
	Alloc       uint64 `json:"alloc"`
	TotalAlloc  uint64 `json:"totalAlloc"`
	Sys         uint64 `json:"sys"`
	HeapAlloc   uint64 `json:"heapAlloc"`
	HeapInuse   uint64 `json:"heapInuse"`
	HeapObjects uint64 `json:"heapObjects"`
	NumGC       uint32 `json:"numGc"`
}

type PprofProfileSummary struct {
	FetchedAt    time.Time           `json:"fetchedAt"`
	ProfileType  string              `json:"profileType"`
	Entries      []PprofProfileEntry `json:"entries"`
	TotalSamples int64               `json:"totalSamples,omitempty"`
	DurationSecs int                 `json:"durationSecs,omitempty"`
}

type PprofProfileEntry struct {
	Name        string  `json:"name"`
	Flat        int64   `json:"flat"`
	FlatPercent float64 `json:"flatPercent"`
	Cum         int64   `json:"cum"`
	CumPercent  float64 `json:"cumPercent"`
}
